### ISSUE TITLE
A U-Net Based Discriminator for Generative Adversarial Networks

### ISSUE THREAD

**Issue by KeisukeShimokawa** [KeisukeShimokawa/pytorch-papers-challenge#17]
### 論文へのリンク

[https://arxiv.org/abs/2002.12655](https://arxiv.org/abs/2002.12655)

### 著者・所属機関

Edgar Schönfeld, Bernt Schiele, Anna Khoreva

### 投稿日時（YYYY-MM-DD）

2020-02-28

### 1. どんなもの？

U-Netベースの構造を採用したDiscriminatorを提案した。DiscriminatorからGeneratorにper-pixelな情報とglobalな情報を送ることで、全体の一貫性と精緻な情報を保った画像を生成することに成功した。また新たにper-pixelな情報を利用した正則化項を導入した。FFHQやCelebAなどのデータセットで、BigGANを上回るFIDを達成した。

**Comment by KeisukeShimokawa** [comment]
とてもシンプルな手法でBigGANよりFIDが優れているのはすごい。

**Comment by KeisukeShimokawa** [comment]
|==|
|--|
|Rs|

**Comment by KeisukeShimokawa** [comment]
**Discriminator Encoder**

|model|image size|
|--|--|
|ResBlock(3, 256) down|h[B, 3, 32, 32] --> h[B, 3, 16, 16]|
|ResBlock(256, 256) down|h[B, 3, 16, 16] --> h[B, 256, 8, 8]|
|ResBlock(256, 256)|h[B, 256, 8, 8] --> h[B, 256, 8, 8]|
|non-Local Block|h[B, 256, 8, 8] --> h[B, 256, 8, 8]|
|ResBlock(256, 256)|h[B, 256, 8, 8] --> h[B, 256, 8, 8]|
|Global Average Pooling|h[B, 256, 8, 8] --> h[B, 256]|
|Linear(256, 1)|h[B, 256] --> h[B, 1]|

**Comment by KeisukeShimokawa** [comment]
**Discriminator Decoder**

このモデルはEncoderの最終層の出力 (`h[B, 256, 8, 8]`) を使用する

|model|image size|
|--|--|
|ResBlock(256, 256)|h[B, 256, 8, 8] --> h[B, 256, 8, 8]|
|ResBlock(256, 256)|h[B, 256, 8, 8] --> h[B, 256, 8, 8]|
|ResBlock(256, 256) up|h[B, 256, 8, 8] --> h[B, 256, 16, 16]|
|ResBlock(256, 256) up|h[B, 256, 16, 16] --> h[B, 256, 32, 32]|
|Conv(256, 1)|h[B, 256, 32, 32] --> h[B, 1, 32, 32]|
|Sigmoid||

**Comment by KeisukeShimokawa** [comment]
32x32の場合の全体像

![image](https://user-images.githubusercontent.com/33191285/84409037-da19c980-ac47-11ea-9d8d-bfa8cba3fae7.png)

**Comment by farhodfm** [comment]
Hi @KeisukeShimokawa! 

Did you implement this method by yourself? 
I am asking because I couldn't find any code release for this paper. 
If so, could you share the code? It would be really helpful. 

Thanks in advance

**Comment by KeisukeShimokawa** [comment]
> Hi @KeisukeShimokawa!
> 
> Did you implement this method by yourself?
> I am asking because I couldn't find any code release for this paper.
> If so, could you share the code? It would be really helpful.
> 
> Thanks in advance

I am in the process of implementing this method. It is a trial-and-error process because there are parts of the paper that are unclear within the paper alone, such as learning methods.

**Comment by farhodfm** [comment]
@KeisukeShimokawa, thank you for your reply.

I see. Could you share the information on which parts of the paper are unclear? 
Maybe I missed those points that are important to understand the paper.

Thank you once again

**Comment by KeisukeShimokawa** [comment]
> @KeisukeShimokawa, thank you for your reply.
> 
> I see. Could you share the information on which parts of the paper are unclear?

Thanks.

I created the pseudo code of UNet-based Discriminator including both CutMix Augmentation and  Consistency Regularization.
Do you think my understanding is correct?

![Algorithm](https://user-images.githubusercontent.com/33191285/86596908-f22a0200-bfd5-11ea-9a3e-42b0f519bfdc.png)

**Comment by tatatawab** [comment]
非常に分かりやすいシェアありがとうございます。

![75741066-c5768d00-5d4c-11ea-9fa6-9ecce5406d98](https://user-images.githubusercontent.com/49480985/90528575-be81f080-e1ad-11ea-9580-0dd875e0e17c.png)


上記のロスについて、ピクセル単位での分類確率の総和(平均)をとることに関してはどう解釈されていますか？
結局のところ、U-net Dの出力map全体の代表値を算出してしまうことに違和感を感じています。
代表値をとってしまうと、同じロス値に収束するような組み合わせ(U-net Dの出力map)はいっぱいありそうです。
その場合、ピクセルごとの判定結果を正しくGにフィードバックできているのかな、と。

https://github.com/KeisukeShimokawa/papers-challenge/blob/4640c006fdbb29a90841a911897c1b4f7a1d92e3/src/gan/UNetDiscriminator/models/unetgan_32.py#L296-L310

**Comment by KeisukeShimokawa** [comment]
式(4)では、Discriminatorが出力する分類確率に対する交差エントロピーでの損失値を計算しています。ですので、分類確率ではなく各ピクセルに対する損失の総和を計算しています。

ご提示していただいている実装部分の通り、`reduction="mean"`でバッチ内の各画像のピクセル単位の損失値の平均値を計算しています。

なお`binary_cross_entropy_with_logits`の実装は、以下の公式コードを参考にしてください。

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp#L202-L218

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp#L19-L28

**Comment by KeisukeShimokawa** [comment]
> @tatatawab さん

式(4)では、Discriminatorが出力する分類確率に対する交差エントロピーでの損失値を計算しています。ですので、分類確率ではなく各ピクセルに対する損失の総和を計算しています。

ご提示していただいている実装部分の通り、`reduction="mean"`でバッチ内の各画像のピクセル単位の損失値の平均値を計算しています。

なお`binary_cross_entropy_with_logits`の実装は、以下の公式コードを参考にしてください。

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp#L202-L218

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp#L19-L28
